File: packages/server/src/domains/flags/repo.ts
import { getOne, getMany, query, withTransaction } from '../../shared/db.js';
import type { Flag, CreateFlagInput, UpdateFlagInput } from '@flick/shared';

export async function findByProject(
  projectId: string,
  opts: { archived?: boolean; tags?: string[] } = {},
): Promise<Flag[]> {
  let sql = 'SELECT * FROM flags WHERE project_id = $1';
  const params: unknown[] = [projectId];
  let idx = 2;

  if (opts.archived !== undefined) {
    sql += ` AND archived = $${idx++}`;
    params.push(opts.archived);
  }
  if (opts.tags && opts.tags.length > 0) {
    sql += ` AND tags && $${idx++}`;
    params.push(opts.tags);
  }

  sql += ' ORDER BY created_at DESC';
  return getMany<Flag>(sql, params);
}

export async function findById(id: string): Promise<Flag | null> {
  return getOne<Flag>('SELECT * FROM flags WHERE id = $1', [id]);
}

export async function findByKey(projectId: string, key: string): Promise<Flag | null> {
  return getOne<Flag>('SELECT * FROM flags WHERE project_id = $1 AND key = $2', [projectId, key]);
}

export async function create(projectId: string, input: CreateFlagInput): Promise<Flag> {
  return withTransaction(async (client) => {
    const { rows: [flag] } = await client.query<Flag>(
      `INSERT INTO flags (project_id, key, name, description, gate_type, tags)
       VALUES ($1, $2, $3, $4, $5, $6) RETURNING *`,
      [projectId, input.key, input.name, input.description ?? '', input.gate_type, input.tags ?? []],
    );

    // Auto-create flag_environments for all project environments
    const { rows: envs } = await client.query(
      'SELECT id FROM environments WHERE project_id = $1',
      [projectId],
    );

    const defaultConfig = input.gate_type === 'percentage'
      ? JSON.stringify({ percentage: 0, sticky: true })
      : '{}';

    for (const env of envs) {
      await client.query(
        `INSERT INTO flag_environments (flag_id, environment_id, enabled, gate_config)
         VALUES ($1, $2, false, $3)`,
        [flag.id, env.id, defaultConfig],
      );
    }

    return flag;
  });
}

export async function update(id: string, input: UpdateFlagInput): Promise<Flag | null> {
  const fields: string[] = [];
  const values: unknown[] = [];
  let idx = 1;

  if (input.name !== undefined) { fields.push(`name = $${idx++}`); values.push(input.name); }
  if (input.description !== undefined) { fields.push(`description = $${idx++}`); values.push(input.description); }
  if (input.tags !== undefined) { fields.push(`tags = $${idx++}`); values.push(input.tags); }

  if (fields.length === 0) return findById(id);

  values.push(id);
  return getOne<Flag>(
    `UPDATE flags SET ${fields.join(', ')} WHERE id = $${idx} RETURNING *`,
    values,
  );
}

export async function archive(id: string): Promise<Flag | null> {
  return getOne<Flag>('UPDATE flags SET archived = true WHERE id = $1 RETURNING *', [id]);
}

export async function restore(id: string): Promise<Flag | null> {
  return getOne<Flag>('UPDATE flags SET archived = false WHERE id = $1 RETURNING *', [id]);
}

export async function remove(id: string): Promise<boolean> {
  const result = await getOne<Flag>('DELETE FROM flags WHERE id = $1 RETURNING id', [id]);
  return result !== null;
}

File: packages/server-rust/src/domains/flags/repo.rs
use chrono::{DateTime, Utc};
use serde::Serialize;
use serde_json::json;
use sqlx::PgPool;
use uuid::Uuid;

use crate::error::AppError;

#[derive(Debug, sqlx::FromRow, Serialize)]
pub struct FlagRow {
    pub id: Uuid,
    pub project_id: Uuid,
    pub key: String,
    pub name: String,
    pub description: String,
    pub gate_type: String,
    pub tags: Vec<String>,
    pub archived: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

pub async fn find_by_project(
    db: &PgPool,
    project_id: Uuid,
    archived: Option<bool>,
    tags: Option<Vec<String>>,
) -> Result<Vec<FlagRow>, sqlx::Error> {
    let mut query = String::from("SELECT * FROM flags WHERE project_id = $1");
    let mut param_idx = 1u32;

    if archived.is_some() {
        param_idx += 1;
        query.push_str(&format!(" AND archived = ${param_idx}"));
    }

    if tags.is_some() {
        param_idx += 1;
        query.push_str(&format!(" AND tags && ${param_idx}"));
    }

    query.push_str(" ORDER BY created_at DESC");

    let mut q = sqlx::query_as::<_, FlagRow>(&query).bind(project_id);

    if let Some(archived) = archived {
        q = q.bind(archived);
    }

    if let Some(tags) = tags {
        q = q.bind(tags);
    }

    q.fetch_all(db).await
}

pub async fn find_by_id(db: &PgPool, id: Uuid) -> Result<Option<FlagRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagRow>("SELECT * FROM flags WHERE id = $1")
        .bind(id)
        .fetch_optional(db)
        .await
}

pub async fn find_by_key(
    db: &PgPool,
    project_id: Uuid,
    key: &str,
) -> Result<Option<FlagRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagRow>(
        "SELECT * FROM flags WHERE project_id = $1 AND key = $2",
    )
    .bind(project_id)
    .bind(key)
    .fetch_optional(db)
    .await
}

pub async fn create(
    db: &PgPool,
    project_id: Uuid,
    key: &str,
    name: &str,
    description: &str,
    gate_type: &str,
    tags: &[String],
) -> Result<FlagRow, AppError> {
    let mut tx = db.begin().await.map_err(|e| AppError::Database(e))?;

    let flag = sqlx::query_as::<_, FlagRow>(
        "INSERT INTO flags (project_id, key, name, description, gate_type, tags)
         VALUES ($1, $2, $3, $4, $5, $6)
         RETURNING *",
    )
    .bind(project_id)
    .bind(key)
    .bind(name)
    .bind(description)
    .bind(gate_type)
    .bind(tags)
    .fetch_one(&mut *tx)
    .await
    .map_err(|e| AppError::Database(e))?;

    // Get all environment ids for this project
    let env_ids: Vec<(Uuid,)> =
        sqlx::query_as("SELECT id FROM environments WHERE project_id = $1")
            .bind(project_id)
            .fetch_all(&mut *tx)
            .await
            .map_err(|e| AppError::Database(e))?;

    // Default gate config based on gate type
    let gate_config = if gate_type == "percentage" {
        json!({"percentage": 0, "sticky": true})
    } else {
        json!({})
    };

    // Insert a flag_environment row for each environment
    for (env_id,) in &env_ids {
        sqlx::query(
            "INSERT INTO flag_environments (flag_id, environment_id, enabled, gate_config)
             VALUES ($1, $2, false, $3)",
        )
        .bind(flag.id)
        .bind(env_id)
        .bind(&gate_config)
        .execute(&mut *tx)
        .await
        .map_err(|e| AppError::Database(e))?;
    }

    tx.commit().await.map_err(|e| AppError::Database(e))?;

    Ok(flag)
}

pub async fn update(
    db: &PgPool,
    id: Uuid,
    name: Option<&str>,
    description: Option<&str>,
    tags: Option<&[String]>,
) -> Result<Option<FlagRow>, sqlx::Error> {
    if name.is_none() && description.is_none() && tags.is_none() {
        return find_by_id(db, id).await;
    }

    let mut query = String::from("UPDATE flags SET updated_at = NOW()");
    let mut param_idx = 0u32;

    if name.is_some() {
        param_idx += 1;
        query.push_str(&format!(", name = ${param_idx}"));
    }

    if description.is_some() {
        param_idx += 1;
        query.push_str(&format!(", description = ${param_idx}"));
    }

    if tags.is_some() {
        param_idx += 1;
        query.push_str(&format!(", tags = ${param_idx}"));
    }

    param_idx += 1;
    query.push_str(&format!(" WHERE id = ${param_idx} RETURNING *"));

    let mut q = sqlx::query_as::<_, FlagRow>(&query);

    if let Some(n) = name {
        q = q.bind(n);
    }

    if let Some(d) = description {
        q = q.bind(d);
    }

    if let Some(t) = tags {
        q = q.bind(t);
    }

    q = q.bind(id);

    q.fetch_optional(db).await
}

pub async fn archive(db: &PgPool, id: Uuid) -> Result<Option<FlagRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagRow>(
        "UPDATE flags SET archived = true, updated_at = NOW() WHERE id = $1 RETURNING *",
    )
    .bind(id)
    .fetch_optional(db)
    .await
}

pub async fn restore(db: &PgPool, id: Uuid) -> Result<Option<FlagRow>, sqlx::Error> {
    sqlx::query_as::<_, FlagRow>(
        "UPDATE flags SET archived = false, updated_at = NOW() WHERE id = $1 RETURNING *",
    )
    .bind(id)
    .fetch_optional(db)
    .await
}

pub async fn remove(db: &PgPool, id: Uuid) -> Result<Option<Uuid>, sqlx::Error> {
    let row: Option<(Uuid,)> =
        sqlx::query_as("DELETE FROM flags WHERE id = $1 RETURNING id")
            .bind(id)
            .fetch_optional(db)
            .await?;
    Ok(row.map(|(id,)| id))
}
